Homework-1/utils/ngram_utils.py

```diff
@@ -0,0 +1,59 @@
+# File: ngram_utils.py
+
+import io
+from collections import Counter
+import nltk
+from nltk.tokenize import word_tokenize
+import math
+
+# Ensure NLTK resources are available
+nltk.download('punkt')
+
+def read_file(file_path):
+    """Read the file content and return as a single string."""
+    with io.open(file_path, 'r', encoding='utf8') as file:
+        return file.read()
+
+def tokenize(text):
+    """Tokenize text using NLTK's word tokenizer."""
+    return word_tokenize(text.lower())
+
+def generate_ngrams(tokens, n):
+    """Generate n-grams from a list of tokens."""
+    ngrams = zip(*[tokens[i:] for i in range(n)])
+    return [" ".join(ngram) for ngram in ngrams]
+
+def get_ngrams_from_file(file_path, n):
+    """Read file, tokenize and generate n-grams."""
+    text = read_file(file_path)
+    tokens = tokenize(text)
+    ngrams = generate_ngrams(tokens, n)
+    return ngrams
+
+def ngram_frequencies(ngrams):
+    """Calculate frequencies of n-grams."""
+    return Counter(ngrams)
+
+def calculate_probability(ngrams, ngram_freqs, total_ngrams, vocabulary_size, smoothing=1):
+    """Calculate probabilities with smoothing for n-grams."""
+    probabilities = {}
+    for ngram in ngrams:
+        count = ngram_freqs.get(ngram, 0)
+        probability = (count + smoothing) / (total_ngrams + smoothing * vocabulary_size)
```

Comment:
This isn't the n-gram model. For the n-gram model, you need to calculate:

$\frac{c(w_i, w_{i+1})}{c(w_i)}$

(adjusting for whatever smoothing you need)

Reply:
E.g. https://github.com/jacob5412/2024SP-MSAI-337/blob/main/Homework-1/models/ngrams/ngrams.py#L74

Reply:
![Screenshot 2024-04-15 at 18 08 22](https://github.com/jacob5412/2024SP-MSAI-337/assets/26154487/4c10dbd5-f3cb-42de-9b64-7dd2eb7f53ca)